static/js/components/VideoPlayer.js

```diff
@@ -1,37 +1,225 @@
+    const { video } = this.props;
+    if (this.player) {
+      // Remove existing tracks for deleted subtitles
+      const tracks = this.player.textTracks();
+      let subtitleUrls = video.videosubtitle_set.map((subtitle: VideoSubtitle) => (makeVideoSubtitleUrl(subtitle)));
+      let trackUrls = [];
+      for (let idx = 0; idx < tracks.length; idx++) {
+        if (this.player && tracks[idx] && !subtitleUrls.includes(tracks[idx].src)) {
+          this.player.removeRemoteTextTrack(tracks[idx]);
+        } else {
+          trackUrls.push(tracks[idx].src);
+        }
       }
+      // Add tracks for any new subtitles associated with the video
+      video.videosubtitle_set.map((subtitle: VideoSubtitle) => {
+        const subUrl = makeVideoSubtitleUrl(subtitle);
+        if (!trackUrls.includes(subUrl)) {
+          this.player.addRemoteTextTrack({
+            kind: "captions",
+            src: subUrl,
+            srcLang: subtitle.language,
+            label: subtitle.language_name
+          });
+        }
+      });
+    }
+  }
+
+  drawCanvas(canvas: HTMLCanvasElement, shiftX: boolean, shiftY: boolean) {
+    function drawCanvasImage(canvas, videoNode, shiftX, shiftY) {
+      const x = shiftX ? Math.floor(videoNode.videoWidth / 2) : 0;
+      const y = shiftY ? Math.floor(videoNode.videoHeight / 2) : 0;
+      const context = canvas.getContext('2d');
+      context.drawImage(
+        videoNode,
+        x, y, Math.floor(videoNode.videoWidth / 2), Math.floor(videoNode.videoHeight / 2),
+        0, 0, canvas.width, canvas.height
+      );
+      setTimeout(drawCanvasImage, 20, canvas, videoNode, shiftX, shiftY);
+    }
+
+    canvas.width = Math.floor(this.videoNode.offsetWidth / 4) - 2;
+    canvas.height = Math.floor(this.videoNode.offsetHeight / 4) - 2;
+    if (canvas && this.videoNode) {
+      drawCanvasImage(canvas, this.videoNode, shiftX, shiftY);
+    }
+  }
+
+  configureCameras() {
+    if (this.cameras) {
+      const canvasElements = this.cameras.getElementsByTagName("canvas");
+      Object.keys(canvases).map(corner => {
+        this.drawCanvas(
+          canvasElements[corner],
+          canvases[corner].shiftX,
+          canvases[corner].shiftY
+        );
+      });
+    }
+  }
+
+  isFullscreen() {
+    // $FlowFixMe
+    return document[FULLSCREEN_API.fullscreenElement];
+  }
+
+  cropVideo() {
+    const { selectedCorner } = this.props;
+    if (selectedCorner) {
+      const shiftX = canvases[selectedCorner].shiftX;
+      const shiftY = canvases[selectedCorner].shiftY;
+      const transformProps = ['transform', 'WebkitTransform', 'MozTransform', 'msTransform', 'OTransform'];
+      let prop = transformProps[0];
+
+      transformProps.some((property: string) => {
+        prop = property;
+        return (typeof this.player.el_.style[prop] !== 'undefined');
+      });
+
+      const aspectRatio = (this.player.videoWidth() / this.player.videoHeight());
+      let videoWidth = Math.min(
```

Comment:
Wouldn't the video width always be the minimum of these values?

Reply:
This prevents black bars from showing up on one side or the other. For example without including `parseInt(window.getComputedStyle(this.videoNode).maxHeight) * aspectRatio`, the following can appear:

 
![screen shot 2017-10-06 at 1 53 02 pm](https://user-images.githubusercontent.com/187676/31291399-bcab8266-aa9d-11e7-92d7-0c4ce7058473.png)